src/common/SelectionList.js

```diff
@@ -0,0 +1,176 @@
+import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';
+import type { NavigationScreenProp } from 'react-navigation';
+
+import type { Dispatch } from '../types';
+import type { ThemeColors } from '../styles';
+import { connect } from '../react-redux';
+import Label from './Label';
+import { IconDone, IconRight } from './Icons';
+import Touchable from './Touchable';
+import { BRAND_COLOR, ThemeContext } from '../styles';
+import { navigateToSelectionListOptionsScreen } from '../nav/navActions';
+import OptionDivider from './OptionDivider';
+import Screen from './Screen';
+
+export type SelectionOptions<K> = Map<K, {| +label: string |}>;
+
+type OptionsScreenProps<K> = $ReadOnly<{|
+  navigation: NavigationScreenProp<{
+    params: {|
+      label: string,
+      options: SelectionOptions<K>,
+      selectedKey: K,
+      onOptionSelect: <K>(key: K) => void,
+    |},
+  }>,
+|}>;
+
+type OptionsScreenState<K> = {|
+  selectedKey: K,
+|};
+
+/**
+ * A selection option list component.
+ *
+ * @prop label - label for selection list.
+ * @prop options - all available options to select from.
+ * @prop selectedKey - key of currently selected option.
+ * @prop onOptionSelect - callback when any option is selected.
+ */
+export class OptionsScreen<K> extends PureComponent<OptionsScreenProps<K>, OptionsScreenState<K>> {
+  constructor(props: OptionsScreenProps<K>) {
+    super(props);
+    this.state = {
+      selectedKey: props.navigation.state.params.selectedKey,
+    };
+  }
+
+  styles = StyleSheet.create({
+    item: {
+      flexDirection: 'row',
+      justifyContent: 'space-between',
+      alignItems: 'center',
+      paddingTop: 12,
+      paddingBottom: 12,
+      paddingLeft: 16,
+      paddingRight: 16,
+    },
+    selectedValue: { color: BRAND_COLOR },
+  });
+
+  /**
+   * This is suppose to be arrow func
+   * But due to bug in flow, this is defined as regular func,
+   * see https://github.com/facebook/flow/issues/8242
+   */
+  handleOptionSelect(selectedKey: K) {
+    const { onOptionSelect } = this.props.navigation.state.params;
+    onOptionSelect(selectedKey);
+    this.setState({ selectedKey });
+  }
+
+  render() {
+    const { label, options } = this.props.navigation.state.params;
+    const { selectedKey } = this.state;
+
+    return (
+      <Screen title={label}>
+        <FlatList
+          ItemSeparatorComponent={OptionDivider}
+          data={Array.from(options)}
+          keyExtractor={item => String(item[0]).toString()}
+          renderItem={({ item, index }) => (
+            <Touchable onPress={() => this.handleOptionSelect(item[0])}>
+              <View style={this.styles.item}>
+                <Label
+                  style={selectedKey === item[0] && this.styles.selectedValue}
+                  text={item[1].label}
+                />
+                {selectedKey === item[0] && <IconDone size={16} color={BRAND_COLOR} />}
+              </View>
+            </Touchable>
```

Comment:
The `item[0]` and `item[1]` business here is kind of messy and opaque. (Though I sure am glad that Flow understands it and successfully provides type-checking!)

Instead, we can replace the `Array.from(options)` with something like
```js
Array.from(options, ([key, option]) => ({ ...option, key }))
```
so that the items have properties `key` and `label`.